xh: add page (#5690)
tldr-pages/tldr: pages/common/xh.md

# xh

> Friendly and fast tool for sending HTTP requests.
> More information: <https://github.com/ducaale/xh>.

- Send a GET request:

`xh {{httpbin.org/get}}`

> - Send a POST request with JSON body (each key specified is added to the top-level JSON object e.g. `{"name": "john", "age": 25}`):

Reviewer: Perhaps this will read better?
```suggestion
- Send a POST request with a JSON body (key-value pairs are added to a top-level JSON object - e.g. `{"name": "john", "age": 25}`):
```

`xh post {{httpbin.org/post}} {{name=john}} {{age:=25}}`

> - Send a GET request with query parameters (e.g. `id=5&sort=true`):

Reviewer: Perhaps this will read better?
```suggestion
- Send a GET request with query parameters (e.g. `param_a=value_a&param_b=value_b`):
```

> `xh get {{httpbin.org/get}} {{id==5}} {{sort==true}}`

Reviewer: Is this a typo? Also, a small naming convention suggestion.

I've never used this command myself, so I don't know the syntax.
```suggestion
`xh get {{httpbin.org/get}} {{parameter_a=value_a}} {{parameter_b=value_b}}`
```

Author: No, that's not, `==` means that's query string
![image](https://user-images.githubusercontent.com/34353640/114308516-981a1800-9ae4-11eb-8fb1-797a96cf1784.png)

# xh

> > XH: Friendly and fast tool for sending HTTP requests.

Reviewer 2: ```suggestion
> Friendly and fast tool for sending HTTP requests.
```

> More information: <https://github.com/ducaale/xh>.

- Send a GET request:

`xh {{httpbin.org/get}}`

> - Send a POST request with JSON body: `{"name": "ahmed", "age": 25}`:

Reviewer 2: ```suggestion
- Send a POST request with JSON body:
```

Author: I think that may be useful to visualize how JSON will look like

Reviewer 2: What about this, then?
```suggestion
- Send a POST request with JSON body: (each key specified is added to the top-level e.g. `{"name": "john", "age": 25}`):
```

> - Send a POST request with JSON body: (each key specified is added to the top-level JSON object e.g. `{"name": "john", "age": 25}`):

Reviewer 2: ```suggestion
- Send a POST request with JSON body (each key specified is added to the top-level JSON object e.g. `{"name": "john", "age": 25}`):
```

> `xh post {{httpbin.org/post}} {{name=john}} {{age:=25}}`

Reviewer 2: Missing colon?
```suggestion
`xh post {{httpbin.org/post}} {{name:=john}} {{age:=25}}`
```

Author: > Missing colon?

No, thats expected, numbers use `:=` syntax

Reviewer 2: Ah, good to know. Thank you.

> - Send a GET request with querystring id=5&sort=true:

Reviewer 2: ```suggestion
- Send a GET request with query parameters:
```

Author: Same as with JSON body example, I think that might be nice to visualize that.

Reviewer 2: ```suggestion
- Send a GET request with query parameters (e.g. `id=5&sort=true`):
```

`xh get {{httpbin.org/get}} {{id==5}} {{sort==true}}`

> - Send a GET request and include a header named x-api-key with value 12345:
> 
> `xh get {{httpbin.org/get}} {{x-api-key:12345}}`

Reviewer 2: ```suggestion
- Send a GET request with a custom header:

`xh get {{httpbin.org/get}} {{header-name:header-value}}`
```

- Send a PUT request and pipe the result to less:

> `xh put {{httpbin.org/put}} {{id:=49}} {{age:=25}} | less`

Reviewer 2: This example doesn't seem particularly useful—anyone familiar with the command line will understand that the stdout of one command may be passed to the stdin of another; `less` is frequently to demonstrate this in tutorials anyway, hence it doesn't strike me as being useful to include.

Any reason for including it?

Author: Hmm, okay, I'll remove that.

> - Download and save to res.json:
> 
> `xh -d {{httpbin.org/json}} -o {{res.json}}`

Reviewer 2: ```suggestion
- Make a GET request and save the response body to a file:

`xh --download {{httpbin.org/json}} --output {{path/to/file}}`
```
Also note that we tend towards using the long options of commands when possible as they're usually self explanatory and invaluable to the beginner.
